### ISSUE TITLE
Support for Munbyn ITPP941

### ISSUE THREAD

**Issue by proski** [michaelrsweet/lprint#36]
I was able to modify `rastertolabel` in CUPS to support Munbyn ITPP941. I verified the code behavior on MacOS against Munbyn's own `rastertolabel` from its MacOS driver. I was able to achieve exactly the same output. I tried changing all options and I was able to figure out how they affect the output.

The code is available here: https://github.com/proski/cups/blob/beeprt/filter/rastertolabel.c
The diff: https://github.com/proski/cups/commit/9ade138db4387ed016f70feb11a3b7a05daf04ca

Beeprt and Rollo label printers are apparently compatible, but I cannot guarantee anything, as I don't have them. I know they share the same closed-source Linux CUPS filter called `rastertolabelbeeprt` (it has some bugs compared to the MacOS version). Munbyn drivers for Windows use "beeprt" in the driver names. So I tried to follow the pattern and used `BEEPRT` as the symbol.

`cupsModelNumber` 20 is used in PPDs of all three devices. While it's not the choice of CUPS developers, I believe it can be accommodated.

There is also a new 300 dpi model on Munbyn, called ITPP988. I could not get its filter run on MacOS without the hardware (it actually needs the PRINTER environment variable to point to the printer name, and rejects the model I have). However, I did my best to write the code that would likely accommodate that printer. The 203 dpi model assumes exactly 8 dots per millimeter, so I assume the 300 dpi model is using 12 dots per millimeter.

Anyway, that code needs to be ported to lprint if it cannot be merged to CUPS.

I'm going to try it, but I would welcome comments about my code for far and ideas how to proceed.

**Comment by michaelrsweet** [comment]
@proski Thanks for the head's up on this. I actually found out that the Munbyn and Beeprt printers are both knockoffs of the Rollo printer. I really don't want to get involved in the (legal) dispute there, so I don't be adding support for those printers to LPrint.

Sorry.

**Comment by proski** [comment]
@michaelrsweet I understand the reasons. Please feel free to contact me if anything changes.

Here's there funny part. Look at the Rollo's download page for Linux: https://www.rollo.com/driver-linux/

Download the Ubuntu driver and look inside. There is a file called `rastertolabelbeeprt`. The PDF file has the Beeprt name redacted with a red line in multiple places. You can see the descended of "p" sticking out. On page 8 (Set Printer Options), the "editor" was too lazy to use a thicker line, so "Beeprt" can be seen between two red lines. Pathetic.

**Comment by proski** [comment]
It turns out the language used by all those printers is called TSPL. It's well documented: https://www.servopack.de/support/tsc/TSPL_TSPL2_Programming.pdf
So maybe the implementation can simply refer to the labels printers that support TSPL?

**Comment by stepanselyuk** [comment]
Hi @proski,
I have the same issue with Munbyn (300dpi) label printer. On x64 Ubuntu I run it using the driver (PPD + rastertolabel binary) provided by Munbyn, but in PPD mentioned only 203dpi resolution. I fixed the PPD file, but I currently have an issue with a white label between prints, seems something is coded inside of rastertolabel binary. For media size 2inch (round label) I've set horizontal print offset as 19mm and vertical to -3mm, it placed print nicely then, but got an issue with a white label between.

But for Raspberry Pi they provide only driver for POS and 32bit only and for ancient kernel 2.6. On RPi I actually installed the driver from Rollo and it somehow works, but with crazy offsets and so on (probably also an issue with 203dpi). 

As I've seen in your code it can orient to what is specified in HwResolution and correct calculate dots per mm. Is it still an option do use that file and compile CUPS manually and probably just replace rastertolabel binary, right? 

Hi @michaelrsweet, so if there is an issue with knockoffs of the Rollo, then probably the code included in LPrint can be called after Rollo name, without using BeePRT or Munbyn names. If these devices are just the same thing from some OEM provider.
